TileDB.CSharp/src/ArrayExtension.cs

```diff
@@ -0,0 +1,360 @@
+        {
+            if (array == null || string.IsNullOrEmpty(key) || string.IsNullOrEmpty(value))
+            {
+                return;
+            }
+            try
+            {
+
+                System.Text.StringBuilder sb = new System.Text.StringBuilder();
+                sb.Append("{");
+
+                sb.AppendFormat("\"key\":\"{0}\"", key);
+
+                sb.AppendFormat(",\"value_type\":{0}", (int)TileDB.DataType.TILEDB_STRING_ASCII);
+
+                sb.AppendFormat(",\"value\":\"{0}\"", value);
+
+                sb.Append("}");
+
+                array.put_metadata_by_json_str_for_key(key, sb.ToString());
+
+            }
+            catch (TileDB.TileDBError tdbe)
+            {
+                System.Console.WriteLine("caught TileDBError:");
+                System.Console.WriteLine(tdbe.Message);
+            }
+            catch (System.Exception e)
+            {
+                System.Console.WriteLine("caught exception:");
+                System.Console.WriteLine(e.Message);
+            }
+        }
+        public static void AddArrayMetadataByStringMap(this TileDB.Array array, System.Collections.Generic.Dictionary<string, string> strmap)
+        {
+            if (array == null || strmap == null || strmap.Count == 0)
+            {
+                return;
+            }
+            try
+            {
+ 
+                System.Text.StringBuilder sb = new System.Text.StringBuilder();
+                sb.Append("{");
+
+                sb.AppendFormat("\"metadata_num\":{0}", strmap.Count);
+                sb.Append(",\"metadata\":[");
+
+                int count = 0;
+                foreach (var item in strmap)
+                {
+                    if (count > 0)
+                    {
+                        sb.Append(",");
+                    }
+
+                    sb.Append("{");
+
+                    sb.AppendFormat("\"key\":\"{0}\"", item.Key);
+
+                    sb.AppendFormat(",\"value_type\":{0}", (int)TileDB.DataType.TILEDB_STRING_ASCII);
+
+                    sb.AppendFormat(",\"value\":\"{0}\"", item.Value);
+
+                    sb.Append("}");
+
+
+                    ++count;
+                }
+
+                sb.Append("]");
+
+                sb.Append("}");
+
+                array.put_metadata_by_json_str(sb.ToString());
+            }
+            catch (TileDB.TileDBError tdbe)
+            {
+                System.Console.WriteLine("caught TileDBError:");
+                System.Console.WriteLine(tdbe.Message);
+            }
+            catch (System.Exception e)
+            {
+                System.Console.WriteLine("caught exception:");
+                System.Console.WriteLine(e.Message);
+            }
+        }
+
+        public static void AddArrayMetadataByList<T>(this TileDB.Array array, string key, System.Collections.Generic.List<T> list) where T : struct
```

Comment:
TileDB supports typed metadata, including multi-values for primitive types, so this seems to be redundant. Am I missing something here? IIUC this is encoding in JSON, storing the string, then decoding again into primitive types again on read. Is there a reason not to put the values as natively typed metadata?

Reply:
Mostly for convenience reason. since metadata and schemadata are usually small, that will not be a performance issue. I will add another implementation in native typed data reading/writing, but would like to keep the json representaion of metadata and array schema.

Reply:
The reason we have native metadata support is so that other API clients can read it. A custom JSON encoding is not compatible with other clients (we do use JSON for some python-specific extended type information in tiledb-py, but that is for attribute interpretation purposes; all other metadata is encoded natively).

Reply:
Json is only an intermediate format to transfer data from c# to c++.  the final format is the same. Other api clients can read it without problem